daemon/image.go

```diff
@@ -113,12 +121,228 @@ func (daemon *Daemon) GetCachedImage(imgID image.ID, config *containertypes.Conf
 	return getMatch(siblings)
 }
 
+type daemonImageCacheForBuild struct {
+	// cacheFromImages here is a map of (provided) names to the actual images it represents
+	cacheFromImages map[string]*image.Image
+	// cacheFromImageHistories also provides a map back to a historyWithSourceT struct
+	cacheFromImageHistories map[string][]historyWithSourceT
+	// daemon stores a reference to the daemon that backs this cache
+	daemon *Daemon
+}
+
+func (daemon *Daemon) MakeImageCacheForBuild(cacheFrom []string) builder.ImageCacheForBuild {
+	cache := &daemonImageCacheForBuild{
+		daemon:                  daemon,
+		cacheFromImages:         make(map[string]*image.Image),
+		cacheFromImageHistories: make(map[string][]historyWithSourceT),
+	}
+
+	// for each cacheFrom image, set up the channels & coroutine for scrolling forward through
+	// its history and comparing it to what's being built
+	for _, cacheFromImageName := range cacheFrom {
+		cacheFromImage, err := daemon.GetImage(cacheFromImageName)
+		if err != nil {
+			logrus.Warnf("Could not look up %s for cache resolution, skipping: %s", cacheFromImageName, err)
+			continue
+		}
+
+		cache.cacheFromImages[cacheFromImageName] = cacheFromImage
+		cache.cacheFromImageHistories[cacheFromImageName] = makeHistoryWithSource(cacheFromImage)
+	}
+
+	return cache
+}
+
+// In the history array, we have pairs of (command, resultingLayerID). What we actually want to be able
+// to compare is pairs of (sourceLayerID, command), and if we have a match, consult resultingLayerID.
+// We also don't directly have source/resultingLayerID, but rather a boolean "did create new layer" flag.
+// Define a struct to store this mapping for convenience.
+type historyWithSourceT struct {
```

Comment:
+1 . One compare function should be enough to compare the history chain and it can use a helper function that finds the corresponding diffID for the `History` object.